stlab/concurrency/future.hpp

```diff
@@ -291,24 +291,24 @@ struct shared_base<T, enable_if_copyable<T>> : std::enable_shared_from_this<shar
     }
 
     template <typename F>
-    auto then_r(bool unique, F f) { return then_r(unique, _executor, std::move(f)); }
+    auto then_r(bool unique, F&& f) { return then_r(unique, _executor, std::forward<F>(f)); }
 
     template <typename S, typename F>
-    auto then_r(bool unique, S s, F f) {
-        return recover_r(unique, std::move(s), [_f = std::move(f)](auto x){
+    auto then_r(bool unique, S&& s, F&& f) {
+        return recover_r(unique, std::forward<S>(s), [_f = std::forward<F>(f)](auto x){
             return _f(std::move(x).get_try().get());
         });
     }
 
     template <typename F>
-    auto recover_r(bool unique, F f) { return recover_r(unique, _executor, std::move(f)); }
+    auto recover_r(bool unique, F&& f) { return recover_r(unique, _executor, std::forward<F>(f)); }
 
     template <typename S, typename F>
-    auto recover_r(bool unique, S s, F f) {
-        if (!unique) return recover(std::move(s),std::move(f));
+    auto recover_r(bool unique, S s, F&& f) {
+        if (!unique) return recover(std::forward<S>(s),std::forward<F>(f));
```

Comment:
Is it correct to use a forward< S > on an l-value?

Reply:
Fixed.

Reply:
The following include in utility.h is misspelled (missing u):
#include <stlab/concurrency/immediate_exector.hpp>